math/check_prime.cpp

```diff
@@ -2,55 +2,81 @@
  * Copyright 2020 @author omkarlanghe
  *
  * @file
- * A simple program to check if the given number if prime or not.
- *
  * @brief
- * Reduced all possibilities of a number which cannot be prime.
- * Eg: No even number, except 2 can be a prime number, hence we will increment
- * our loop with i+6 jumping and check for i or i+2 to be a factor of the
- * number; if it's a factor then we will return false otherwise true after the
- * loop terminates at the terminating condition which is (i*i<=num)
+ * A simple program to check if the given number if prime or not.
+ * https://en.wikipedia.org/wiki/Primality_test
+ * @details
+ * additional ways to solve the prime check problem
+ * https://cp-algorithms.com/algebra/primality_tests.html#practice-problems
  */
 
 #include <cassert>   /// for assert
 #include <iostream>  /// for IO operations
 
 /**
- * Function to check if the given number is prime or not.
+ * @brief Function to check if the given number is prime or not.
  * @param num number to be checked.
- * @return if number is prime, it returns @ true, else it returns @ false.
+ * @return true if number is a prime
+ * @return false if number is not a prime.
  */
 template <typename T>
 bool is_prime(T num) {
-    bool result = true;
+    /*!
+     * Reduce all possibilities of a number which cannot be prime with the first
+     * 3 if, else if conditionals. Example: Since no even number, except 2 can
+     * be a prime number and the next prime we find after our checks is 5,
+     * we will start the for loop with i = 5. then for each loop we increment
+     * i by +6 and check if i or i+2 is a factor of the number; if it's a factor
+     * then we will return false. otherwise, true will be returned after the
+     * loop terminates at the terminating condition which is i*i <= num
+     */
     if (num <= 1) {
         return false;
     } else if (num == 2 || num == 3) {
         return true;
-    } else if ((num % 2) == 0 || num % 3 == 0) {
+    } else if (num % 2 == 0 || num % 3 == 0) {
         return false;
     } else {
-        for (T i = 5; (i * i) <= (num); i = (i + 6)) {
-            if ((num % i) == 0 || (num % (i + 2) == 0)) {
-                result = false;
-                break;
+        for (T i = 5; i * i <= num; i = i + 6) {
+            if (num % i == 0 || num % (i + 2) == 0) {
+                return false;
             }
         }
     }
-    return (result);
+    return true;
 }
 
 /**
- * Main function
+ * @brief Self-test implementations
+ * @returns void
  */
-int main() {
-    // perform self-test
+static void tests() {
+    /*!
+     * each assert test checks if the is_prime function returns the expected
+     * boolean value to whether its argument is prime
+     */
+    assert(is_prime(1) == false);
+    assert(is_prime(2) == true);
+    assert(is_prime(3) == true);
+    assert(is_prime(4) == false);
+    assert(is_prime(4) == false);
+    assert(is_prime(7) == true);
+    assert(is_prime(19) == true);
```

Comment:
Since the function is generic i think it might be a good idea to test numbers which are not just integers; if they fail i think its better to make the function non generic

Reply:
I was confused by the use of a template in this algo, since prime numbers are natural numbers only, they can't be floats. Is there any trick I'm missing with using a template here? or can I just change it to long long (int64).

Reply:
yes removing the templates would be fine